man/qrdr.8
.\" $Header$
.\"
.\"  Man page for FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993
.\"
.fp 4 CW
.ds ]W Funet-Nje 8 UNIX 3.0
.TH QRDR 8
.SH NAME
qrdr \- incoming bitnet queue peeking tool
.SH SYNOPSIS
.IP qrdr 0.5i 0
.RB [ " \-u \fIuser\fR " ]
.RB [ " \-n " ]
.RB [ " \-l " ]
.RI [ " dirpaths\fR,\fI filepaths " ]
.SH DESCRIPTION
.IX "qrdr"
.IX "bitnet"
.IX "funetnje"
.IX "mailify"
.IX "receive"
.LP
.I Qrdr
gives users a tool to see what they have in their incoming BITNET spool
queues.
Depending on protection it also lets others to see what user XYZ has in
his/her queue, or ANY directory possibly containing BITNET spool files,
or just any specified files...
.LP
When
.RB ` "\-u \fIuser" '
has not been used, default behaviour of
.I qrdr
is to find out who user is making query  by asking `whoami',
eg getting value of
.B LOGNAME
environment variable.
(In SysV this is
.B USER
variable.)
If environment doesn't give answer,
.I /etc/passwd
file is consulted with
effective uid.
.LP
Spool directory is determined using same mechanism as in
.I funetnje
main program.  Sample configuration of file procesing exits gives directory
.RI ` /var/spool/bitspool '
in which UPPERCASE userids (or pseudoids - MAILER is one such) are `user'
specific subdirectories containing individual spooled files.
.LP
Option
.BR ` \-l '
gives file paths to spool files, and some heuristics of file contents
for
.IR rdrlist (1)
program.
For program-parsing the
.BR ` \-l '
option uses singular
tabulator to separate output fields.
.LP
Example of
.I qrdr
output for user 
.RB ` mea ':
.LP
.nf
.fz 4 -2
\f4QRDR: Spool dir: `/var/spool/bitspool/MEA/'
\f4From:             To:               FName:   FExt:    Type  Form:    SpoolID
\f4NETSERV@FINHUT    MEA@FINFILES      NETSERV  NOTE     PUN A QUFINHUT 0002
\f4NETSERV@FINHUT    MEA@FINFILES      NETSERV  NOTE     PUN A QUFINHUT 0004
\f4MEA@FINFILES      MEA@FINFILES      STANDARD INPUT    PUN C STANDARD 0006
\f4NETSERV@FINHUT    MEA@FINFILES      NETSERV  NOTE     PUN A QUFINHUT 0003
\f4ZEHAVI@ISRAEARN   MEA@FIRIEN        ZEHAVI   NOTE     PUN A STANDARD 0005
\f4...
\f4LISTSERV@SEARN    MEA@FINFILES      INFO     FILELIST PUN N QULISTSV 0023
\f4LISTSERV@FINHUTC  MEA@FINFILES      LISTOWNR REFCARD  PUN N QULISTSV 0024
\f4CMS1@ALIJKU11     MEA@FINFILES      CMS1     NOTE     PUN A STD      0025
\f4Found 23 files.
.fi
.LP
For the
.B \-l
option, recognized data types are:
NETDATA, VMSDUMP, CARD (Cornell CARD), and PUNCH when none of
the previous ones match to input punch records.
For three PRINT formats there are:
PRINT (Machine Carriage Control), PASA (ANSI Carriage Control), and
PCPDS for rare 3rd alternative.
.SH FILES
.nf
.I /etc/funetnje.cf
.I /etc/funetnje.pid
.RI "File processing exit config file, which is defined in " funetnje.cf
.fi
.SH ENVIRONMENT
.IP LOGNAME 1.0i 0
Primary way to find who I am. (BSD)
.IP USER 1.0i 0
Secondary way to find who I am. (SysV)
.SH SEE ALSO
.BR funetnje (8),
.BR rdrlist (1),
.BR receive (1),
.BR mailify (8),
.BR send (1),
.BR sendfile (1),
.BR bmail (8),
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, man-pages by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

qrdr.c
/*----------------------------------------------------------------------*
 |		QRDR		Query user incoming queue		|
 |	For FUNET-NJE by Matti Aarnio  <mea@nic.funet.fi>		|
 |									|
 |  Options:								|
 |	-u user		(default is invoker)				|
 |	-n		(give return code != 0 if pending files)	|
 |      -l		(list original filenames in front if other data,|
 |                       and file arrival date (UNIX ctime) to spool.)  |
 |	dirpath		(point to directory to search)			|
 |	filepath	(point to file to search)			|
 |									|
 |  Default:	Show short listing of files in my (invoker) rdr queue	|
 |									|
 |									|
 *----------------------------------------------------------------------*/

#include <time.h>
#include "consts.h" /* DIR & DIRENTTYPE definitions.. */
#include "prototypes.h"
#include <ctype.h>
#include "clientutils.h"

char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";

int LogLevel = 2;
FILE *LogFd = NULL;

extern char DefaultSpoolDir[256];
extern char DefaultPOSTMASTdir[256];

extern	int	dump_header __(( char const *path, char const *dirpath, const time_t mtime ));
extern	void	usage __(( void ));

char	WhoAmI[10] = "";
char	WhoAmIl[10];	/* Lowercase copy.. */
char	StudyDir[1024];
char	delayprint[1024];
char	*pname = "<QRDR>";

int	BeQuiet = 0;	/* -n gives only non-zero RC if files in queue */
int	BeQuiet2 = 0;	/* -N gives non-zero RC if files in queue, and
			   prints `You have (no) bitnet files in queue' */
int	FilesFound = 0;
int	LongList = 0;	/* -l prints also filenames in front of other data */
int	Debug = 0;

struct namedate {
	char	*name;
	time_t	mtime;
};

static int cmpnamedate(p1,p2)
const struct namedate *p1, *p2;
{
  return (p1->mtime) - (p2->mtime);
}

static int
study_dir(dirpath)
char *dirpath;
{
	DIR	*dirfile;
	DIRENTTYPE *dirp;
	struct stat fstat;
	char curdir[MAXPATHLEN];
	struct namedate *namedates = NULL;
	int entryspace = 0;
	int entrycount = 0;
	int rc;

#if	defined(USG)
	getcwd(curdir,sizeof(curdir));
#else
	getwd(curdir);
#endif
	dirfile = opendir(dirpath);
	if (dirfile == NULL) {
	  if (!BeQuiet && !BeQuiet2) {
	    if (*delayprint)
	      puts(delayprint);
	    *delayprint = 0;
	    printf("QRDR: Can't open directory `%s'\n",dirpath);
	  }
	  return 0;
	}
	rc = 0;
	chdir(dirpath);
	for (dirp = readdir(dirfile);
	     dirp != NULL;
	     dirp = readdir(dirfile)) {
	  if (stat(dirp->d_name,&fstat) != 0)
	    continue;
	  /* printf("QRDR: dirfile: `%s'\n",dirp->d_name); */
	  /* if (strcmp(dirp->d_name,".spoolid")== 0)
	     continue; */ /* old, obsolete.. */
	  if ((fstat.st_mode & S_IFMT) == S_IFREG) {
	    if (entryspace == 0) {
	      entryspace  = 8;
	      namedates = (struct namedate*)malloc(sizeof(struct namedate) *
						   entryspace);
	    } else if (entrycount >= entryspace) {
	      entryspace += 8;
	      namedates = (struct namedate*)realloc(namedates,
						    sizeof(struct namedate) *
						    entryspace);
	    }
	    namedates[entrycount].name = strdup(dirp->d_name);
	    namedates[entrycount].mtime = fstat.st_mtime;
	    ++entrycount;
	  }
	}
	if (entrycount > 0) {
	  int i;
	  qsort(namedates,entrycount,sizeof(struct namedate),cmpnamedate);
	  for (i = 0; i < entrycount; ++i) {
	    rc |= dump_header(namedates[i].name, dirpath, namedates[i].mtime);
	    free(namedates[i].name);
	  }
	  free(namedates);
	}
	closedir(dirfile);
	chdir(curdir);
	if (BeQuiet2 & FilesFound) {
	  printf("You have BITNET files waiting! Use  qrdr -l  to check them!\n");
	  exit(1);	/* Signal founding files... */
	}
	if (BeQuiet & FilesFound)
	  exit(1);	/* Signal founding files... */
	return rc;
}


int
main(argc,argv)
     int argc;
     char *argv[];
{
	struct passwd *passwdent;
	char *p;
	int c;
	char HomeDir[256];
	struct stat fstat;
	int UseDefault = 0;
	extern char *optarg;
	extern int optind;

	pname = *argv;

	*WhoAmI = 0;
	*StudyDir = 0;

	read_configuration();

	/* Process options here! */

	while ((c = getopt(argc, argv, "u:DnNl?")) != -1) {
	  switch (c) {
	    case 'u':
		strncpy(WhoAmI,optarg,8);
		break;
	    case 'D':
		Debug = 1;
		break;
	    case 'N':
		BeQuiet2 = 1;
		BeQuiet = 1;
		break;
	    case 'n':
		BeQuiet = 1;
		break;
	    case 'l':
		LongList = 1;
		break;
	    case '?':
	        usage();
		break;
	    default:
	        usage();
		break;
	  }
	}

	argv += optind;
	argc -= optind;

	if (*WhoAmI == 0) {
	  WhoAmI[8] = 0;
	  p = getenv("LOGNAME");  /* BSD/SunOS/... alike */
	  if (!p) {
	    p = getenv("USER");	/* SysV */
	    if (!p) {
	      passwdent = getpwuid(getuid());
	      if (passwdent)
		strncpy(WhoAmI,passwdent->pw_name,8);
	      else {
		fprintf(stderr,"QRDR: Can't figure out who I am.\n");
		exit(2);
	      }
	    } else
	      strncpy(WhoAmI,p,8);
	  } else
	    strncpy(WhoAmI,p,8);
	}
	strcpy(WhoAmIl,WhoAmI);
	upperstr(WhoAmI);
	    
	if (*StudyDir == 0) {
	  strcpy(StudyDir,DefaultSpoolDir);
	  if ((passwdent = getpwnam( WhoAmIl )) != NULL) {
	    strcpy(HomeDir,passwdent->pw_dir);
	  } else {
	    if (!(BeQuiet || LongList))
	    logger(1,"QRDR: Can't figure home dir of `%s' -- PseudoUser ?\n",
		   WhoAmI);
	    strcpy(HomeDir,"/");
	  }
	  ExpandHomeDir(DefaultSpoolDir,HomeDir,WhoAmI,StudyDir);
	  UseDefault = 1;
	}
	
	*delayprint = 0;
	if (!(BeQuiet || LongList))
	  sprintf(delayprint,"QRDR: Spool dir: `%s'",StudyDir);

	if (!*argv)
	  study_dir(StudyDir);
	else
	  while (*argv) {
	    if (stat(*argv,&fstat) != 0) {
	      if (!UseDefault) { /* Default: Don't worry.. */
		if (*delayprint)
		  puts(delayprint);
		*delayprint = 0;
		printf("QRDR: `%s' is nonexistent!  Can't stat() it.\n",*argv);
	      }
	      ++argv;
	      continue;			/* Bad name ? Skip it */
	    }
	    if ((fstat.st_mode & S_IFMT) == S_IFREG)
	      dump_header(*argv,"",fstat.st_mtime);
	    else if ((fstat.st_mode & S_IFMT) == S_IFDIR)
	      study_dir(*argv);
	    else {
	      if (*delayprint)
		puts(delayprint);
	      *delayprint = 0;
	      printf("QRDR: `%s' Not regular file, or directory!\n",*argv);
	      ++argv;
	      continue;
	    }
	    ++argv;
	  }
	if (BeQuiet2) {
	  if (FilesFound == 0)
	    printf("QRDR: You have no BITNET files waiting.\n");
	  else
	    printf("QRDR: Found %d BITNET file%s waiting for you.\n",
		   FilesFound,(FilesFound==1) ? "":"s" );
	  return (FilesFound != 0);
	}

	if (BeQuiet)
	  return (FilesFound != 0);	/* Signal founding files... */

	if (FilesFound > 0 && !LongList) {
	  if (*delayprint)
	    puts(delayprint);
	  *delayprint = 0;
	  printf("Found %d file%s.\n",FilesFound,(FilesFound==1) ? "":"s" );
	} else if (!LongList) {
	  if (*delayprint)
	    puts(delayprint);
	  *delayprint = 0;
	  printf("No files found.\n");
	}
	return 0;
}

void
Type2Str(type,str)
     const int type;
     char *str;
{
	if (type & F_PRINT)
	  strcpy(str,"PRT");
	else if (type & F_JOB)
	  strcpy(str,"JOB");
	else
	  strcpy(str,"PUN");
}

static	unsigned char INMR01[7] = {0xE0,0xC9,0xD5,0xD4,0xD9,0xF0,0xF1};
static	unsigned char CORNELLCARD[8] =
	{ 0x7A,0xC3,0xC1,0xD9,0xC4,0x40,0x40,0x40 };
static  unsigned char VMSDUMP[12] =
	{ 0xFF,0x0D,0x00,'V','M','S','D','U','M','P',' ','V' };

int
dump_header(path,dirpath,mtime)
     char const *path;
     char const *dirpath;
     const time_t mtime;
{
	int	rc;
	char	FileName[20], FileExt[20], From[20], To[20], Class;
	char	Form[20];
	char	TypeStr[20];
	char	Tag[140];
	char	*ContentsType;
	int	Type, FileId, Format;
	FILE	*fd;
	static	int	once = 1;
	int	FilePosition;
	unsigned char uline[1000];
	char	const *SpoolID = strrchr(path,'/');

	if (SpoolID) ++SpoolID;
	else	       SpoolID = path; 

	if (!(fd = fopen(path,"r"))) return -1;

	rc = parse_header(fd, From, To, FileName, FileExt, &Type,
			  &Class, Form, &Format, Tag, &FileId);

	if (Debug) printf("parse_header(%s) rc=%d\n",path,rc);

	FilePosition = ftell(fd);	/* Save the beginning of real message*/

	/* Lets see if it is NETDATA... */
	while (1) {
	  if (Uread(uline,sizeof uline,fd) < 1) break;
	  if (*uline == 0x80) break; /* PUNCH.. */
	  if (*uline == 0x90) break; /* PRINT */
	  if (*uline == 0xA0) break; /* PASA  */
	  if (*uline == 0xB0) break; /* PCPDS */
	}

	fseek(fd, FilePosition, 0);  /* Get back to beginning of real text */
	ContentsType = "*FAULT*";
	if (*uline == 0x80) {	/* virtual card record */
	  if (memcmp(uline+3,INMR01,7)==0) {
	    ContentsType = "NETDATA";
	  } else if (memcmp(uline+2,VMSDUMP,12)==0) {
	    ContentsType = "VMSDUMP";
	  } else if (memcmp(uline+2,CORNELLCARD,8)==0) {
	    ContentsType = "CARD";
	  } else
	    ContentsType = "PUNCH";
	} else if (*uline == 0x90) { /* PRINTs.. */
	  ContentsType = "PRINT";
	} else if (*uline == 0xA0) { /* PRINTs.. */
	  ContentsType = "PASA";
	} else if (*uline == 0xB0) { /* PRINTs.. */
	  ContentsType = "PCPDS";
	}

	fclose(fd);

	if (rc != 0) {
	  if (*delayprint)
	    puts(delayprint);
	  *delayprint = 0;
	  printf("File: `%s' is NOT FUNET-NJE Spool file!\n",path);
	  return 1;
	}

	FilesFound++;

	Type2Str(Type,TypeStr);

	if (BeQuiet)
	  return 0;

	if (SpoolID) {
	  if (!(SpoolID[0] >= '0' && SpoolID[0] <= '9' &&
		SpoolID[1] >= '0' && SpoolID[1] <= '9' &&
		SpoolID[2] >= '0' && SpoolID[2] <= '9' &&
		SpoolID[3] >= '0' && SpoolID[3] <= '9' &&
		SpoolID[4] == 0))
	    SpoolID = "??";
	}

	/* Print header line, if no long-list.. */
	if (once && !LongList) {
	  if (*delayprint)
	    puts(delayprint);
	  *delayprint = 0;
	  printf("%-17s %-17s %-8s %-8s %4s  %-8s %s\n",
		 "From:","To:","FName:","FExt:","Type","Form:",
		 "SpoolID");
	  once = 0;
	}
	if (*delayprint)
	  puts(delayprint);
	*delayprint = 0;
	if (LongList) {
	  int len = strlen(dirpath);
	  char last = len ? dirpath[len-1] : 0;
	  char cdatestr[20];
	  struct tm *tp = gmtime(&mtime);
	  sprintf(cdatestr,"%04d%02d%02d%02d%02d%02d",
		  tp->tm_year+1900,tp->tm_mon+1,tp->tm_mday,
		  tp->tm_hour,tp->tm_min,tp->tm_sec);
	  printf("%s%s%s\t%s\t%s\t%s\t%s\t%s\t%s\t%c\t%s\t%s\t%s\n",
		 dirpath, (last=='/') ? "":"/", path,
		 ContentsType,
		 From,To,FileName,FileExt,TypeStr,Class,Form,SpoolID,
		 cdatestr);
	} else
	  printf("%-17s %-17s %-8.8s %-8.8s %-3s %c %-8.8s %s\n",
		 From,To,FileName,FileExt,TypeStr,Class,Form,SpoolID);

	return 0;
}

volatile void
bug_check(str)
     char const *str;
{
	logger(1,"QRDR: BUG_CHECK: %s",str);
	exit(100);
}

void
usage()
{
	fprintf(stderr,"\
QRDR:  Query FUNET-NJE readers and/or files in it.\n\
       -u user (default: invoker)\n\
       -n      (give !=0 rc if files in queue dir.)\n\
       dirpath (Check pointed directory)\n\
       filepath (study pointed file)\n");
	fflush(stderr);
}
